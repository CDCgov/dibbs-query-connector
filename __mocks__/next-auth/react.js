const useSession = jest
  .fn()
  .mockReturnValue({ data: undefined, status: "loading" });

const signOut = jest.fn();
/**
 *
 * @param root0 - root0
 * @param root0.children - root0.children
 * @returns - JSX.Element
 */
const SessionProvider = ({ children }) => <>{children}</>;

export { useSession, signOut, SessionProvider };
