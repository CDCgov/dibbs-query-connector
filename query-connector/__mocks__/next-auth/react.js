/* eslint-disable jsdoc/require-returns, jsdoc/require-param-description */
// __mocks__/next-auth/react.js
const useSession = jest.fn().mockReturnValue({ data: undefined });

/**
 *
 * @param root0
 * @param root0.children
 */
const SessionProvider = ({ children }) => <>{children}</>;

export { useSession, SessionProvider };
