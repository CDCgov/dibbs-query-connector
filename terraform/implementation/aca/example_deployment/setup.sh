#!/bin/bash

# If you're running this code from an M-series processor or other ARM variant, you may need to specify the container platform manually.
export DOCKER_DEFAULT_PLATFORM=linux/amd64

# This section provides an example of targeted deployments, if required for your installation.
terraform apply -target=module.foundations -target=module.networking
terraform apply -target=module.container_apps