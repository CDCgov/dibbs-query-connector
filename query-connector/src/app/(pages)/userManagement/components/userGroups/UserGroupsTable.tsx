"use client";

import { useContext, useEffect, useState } from "react";
import classNames from "classnames";
import { Button } from "@trussworks/react-uswds";

import Table from "../../../../ui/designSystem/table/Table";
import { UserManagementContext } from "../UserManagementProvider";
import {
  UserGroup,
  UserRole,
  User,
  UserGroupMembership,
} from "../../../../models/entities/users";
import { CustomUserQuery } from "@/app/models/entities/query";
import styles from "../usersTable/usersTable.module.scss";
import { getContextRole } from "../../utils";
import { getCustomQueries } from "@/app/backend/query-building";

type UserGroupsTableProps = {
  userGroups: UserGroup[];
  fetchGroupMembers: (groupId: string) => Promise<User[]>;
  fetchGroupQueries: (groupId: string) => Promise<CustomUserQuery[]>;
};

/**
 * User groups section in the user management page
 * @param root0 - The properties object
 * @param root0.userGroups The list of user groups to display
 * @param root0.fetchGroupMembers Function to retrieve a group's list of users
 * @param root0.fetchGroupQueries Function to retrieve a group's list of assigned queries
 * @returns The user groups table
 */
const UserGroupsTable: React.FC<UserGroupsTableProps> = ({
  userGroups,
  fetchGroupQueries,
  fetchGroupMembers,
}) => {
  const { openEditSection } = useContext(UserManagementContext);
  const role = getContextRole();

  const [queries, setQueries] = useState<CustomUserQuery[]>([]);

  useEffect(() => {
    async function fetchAllQueries() {
      const queriesResponse = await getCustomQueries();
      setQueries(queriesResponse);
    }
    fetchAllQueries();
  }, []);

  function getMemberLabel(memberSize: number): string {
    return memberSize == 1 ? `${memberSize} member` : `${memberSize} members`;
  }

  function getQueryLabel(querySize: number): string {
    return querySize == 1 ? `${querySize} query` : `${querySize} queries`;
  }

  function renderGroupsTable() {
    if (userGroups.length === 0) {
      return (
        <tr>
          <td colSpan={3}>No user groups found</td>
        </tr>
      );
    }

    function addGroupAssignments(
      query: CustomUserQuery,
      group: UserGroup,
      groupQueries: CustomUserQuery[],
    ) {
      let memberships: UserGroupMembership[] = [];

      groupQueries.forEach((gq) => {
        if (gq.query_id == query.query_id) {
          const membership = {
            membership_id:
              gq.groupAssignments?.find((g) => g.membership_id)
                ?.membership_id || "",
            usergroup_id: group.id,
            usergroup_name: group.name,
            is_member: true,
          };
          memberships.push(membership);
        }
      });
      query.groupAssignments = memberships;
    }

    return userGroups.map((group: UserGroup, idx: number) => (
      <tr key={group.id}>
        <td width={270}>{group.name}</td>
        <td>
          {role != UserRole.SUPER_ADMIN && group.member_size <= 0 ? (
            getMemberLabel(group.member_size)
          ) : (
            <Button
              type="button"
              data-testid={`edit-member-list-${idx}`}
              className={classNames(styles.drawerButton, "text-no-underline")}
              unstyled
              aria-description={`Edit ${group.name} members`}
              onClick={async () => {
                await fetchGroupMembers(group.id).then((members) => {
                  openEditSection(
                    group.name,
                    "Members",
                    "Members",
                    group.id,
                    members as User[],
                  );
                });
              }}
            >
              {getMemberLabel(group.member_size)}
            </Button>
          )}
        </td>
        <td>
          <Button
            type="button"
            data-testid={`edit-query-list-${idx}`}
            className={classNames("text-no-underline")}
            unstyled
            aria-description={`Edit ${group.name} queries`}
            onClick={async () => {
              await fetchGroupQueries(group.id).then((groupQueries) => {
                queries.forEach((query) =>
                  addGroupAssignments(query, group, groupQueries),
                );

                openEditSection(
                  group.name,
                  "Assigned Queries",
                  "Queries",
                  group.id,
                  queries,
                );
              });
            }}
          >
            {getQueryLabel(group.query_size)}
          </Button>
        </td>
      </tr>
    ));
  }

  return (
    <Table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Members</th>
          <th>Assigned queries</th>
        </tr>
      </thead>
      <tbody>{renderGroupsTable()}</tbody>
    </Table>
  );
};

export default UserGroupsTable;
