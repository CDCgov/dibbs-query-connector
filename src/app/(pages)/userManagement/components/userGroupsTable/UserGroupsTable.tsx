"use client";

import { useContext } from "react";
import classNames from "classnames";
import { Button } from "@trussworks/react-uswds";
import { Icon } from "@trussworks/react-uswds";
import Table from "../../../../ui/designSystem/table/Table";
import { UserManagementContext } from "../UserManagementProvider";
import { UserGroup, UserRole, User } from "../../../../models/entities/users";
import { CustomUserQuery } from "@/app/models/entities/query";
import styles from "../../userManagement.module.scss";
import { getRole } from "../../utils";

import { UserManagementMode } from "../../utils";

type UserGroupsTableProps = {
  userGroups: UserGroup[];
  fetchGroupMembers: (groupId: string) => Promise<User[]>;
  fetchGroupQueries: (groupId: string) => Promise<CustomUserQuery[]>;
  openModal: (mode: UserManagementMode, data?: UserGroup | User) => void;
};

/**
 * User groups section in the user management page
 * @param root0 - The properties object
 * @param root0.userGroups The list of user groups to display
 * @param root0.fetchGroupMembers Function to retrieve a group's list of users
 * @param root0.fetchGroupQueries Function to retrieve a group's list of assigned queries
 * @param root0.openModal Function to retrieve a group's list of assigned queries
 * @returns The user groups table
 */
const UserGroupsTable: React.FC<UserGroupsTableProps> = ({
  userGroups,
  fetchGroupQueries,
  fetchGroupMembers,
  openModal,
}) => {
  const { openEditSection } = useContext(UserManagementContext);
  const role = getRole();

  function getMemberLabel(memberSize: number): string {
    return memberSize == 1 ? `${memberSize} member` : `${memberSize} members`;
  }

  function getQueryLabel(querySize: number): string {
    return querySize == 1 ? `${querySize} query` : `${querySize} queries`;
  }

  function renderGroupsTable() {
    if (userGroups.length === 0) {
      return (
        <tr>
          <td colSpan={3}>No user groups found</td>
        </tr>
      );
    }

    return userGroups.map((group: UserGroup, idx: number) => (
      <tr key={group.id} className={styles.userGroupRow}>
        <td width={270}>{group.name}</td>
        <td>
          {role != UserRole.SUPER_ADMIN && group.memberSize <= 0 ? (
            getMemberLabel(group.memberSize)
          ) : (
            <Button
              type="button"
              data-testid={`edit-member-list-${idx}`}
              className={classNames(styles.drawerButton, "text-no-underline")}
              unstyled
              aria-description={`Edit ${group.name} members`}
              onClick={async () => {
                await fetchGroupMembers(group.id).then((members) => {
                  openEditSection(
                    group.name,
                    "Members",
                    "Members",
                    group.id,
                    members,
                  );
                });
              }}
            >
              {getMemberLabel(group.memberSize)}
            </Button>
          )}
        </td>
        <td>
          <Button
            type="button"
            data-testid={`edit-query-list-${idx}`}
            className={classNames("text-no-underline")}
            unstyled
            aria-description={`Edit ${group.name} queries`}
            onClick={async () => {
              await fetchGroupQueries(group.id).then((groupQueries) => {
                openEditSection(
                  group.name,
                  "Assigned Queries",
                  "Queries",
                  group.id,
                  groupQueries,
                );
              });
            }}
          >
            {getQueryLabel(group.querySize)}
          </Button>
        </td>
        {role === UserRole.SUPER_ADMIN ? (
          <td className={styles.actionButtons}>
            <Button
              type="button"
              className="usa-button--unstyled text-bold text-no-underline"
              onClick={() => openModal("edit-group", group)}
            >
              <span className="icon-text padding-right-4 display-flex flex-align-center">
                <Icon.Edit
                  className="height-3 width-3"
                  aria-label="Pencil icon indicating edit ability"
                />
                <span
                  data-testid={`edit-group-${group.id}`}
                  className="padding-left-05"
                >
                  Edit
                </span>
              </span>
            </Button>
            <Button
              type="button"
              className="usa-button--unstyled text-bold text-no-underline"
              onClick={() => {
                openModal("remove-group", group);
              }}
            >
              <span className="icon-text padding-right-4 display-flex flex-align-center">
                <Icon.Delete
                  className="height-3 width-3"
                  aria-label="Trash icon indicating deletion of disease"
                />
                <span className="padding-left-05">Delete</span>
              </span>
            </Button>
          </td>
        ) : (
          <td></td>
        )}
      </tr>
    ));
  }

  return (
    <Table className={styles.userGroupsTable}>
      <thead>
        <tr>
          <th>Name</th>
          <th>Members</th>
          <th>Assigned queries</th>
        </tr>
      </thead>
      <tbody>{renderGroupsTable()}</tbody>
    </Table>
  );
};

export default UserGroupsTable;
