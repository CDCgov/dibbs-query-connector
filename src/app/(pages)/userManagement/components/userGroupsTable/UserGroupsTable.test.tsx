import React from "react";
import { waitFor, screen, render } from "@testing-library/react";
import UserGroupsTable from "./UserGroupsTable";
import {
  mockGroupMany,
  mockGroupWithSingleQuery,
  allGroups,
} from "../../test-utils";
import { RootProviderMock } from "@/app/tests/unit/setup";

jest.mock("next-auth/react");

jest.mock(
  "@/app/ui/components/withAuth/WithAuth",
  () =>
    ({ children }: React.PropsWithChildren) => <div>{children}</div>,
);

jest.mock("@/app/backend/query-building/service", () => ({
  getCustomQueries: jest.fn(),
}));

jest.mock("@/app/backend/user-management", () => ({
  getGroupMembers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getGroupQueries: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUsers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUserGroups: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
}));

describe("User Groups table", () => {
  it("renders the correct label syntax based on member/query size", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UserGroupsTable
          openModal={jest.fn()}
          userGroups={allGroups}
          fetchGroupMembers={jest.fn()}
          fetchGroupQueries={jest.fn()}
          modalData={""}
        />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    const singleMemberBtn = screen.getByTestId(
      `edit-member-list-${allGroups.indexOf(mockGroupWithSingleQuery)}`,
    );
    const multiMemberBtn = screen.getByTestId(
      `edit-member-list-${allGroups.indexOf(mockGroupMany)}`,
    );
    const singleQueryBtn = screen.getByTestId(
      `edit-query-list-${allGroups.indexOf(mockGroupWithSingleQuery)}`,
    );
    const multiQueryBtn = screen.getByTestId(
      `edit-query-list-${allGroups.indexOf(mockGroupMany)}`,
    );

    expect(singleMemberBtn).toHaveTextContent(
      mockGroupWithSingleQuery.memberSize + " member",
    );

    expect(multiMemberBtn).toHaveTextContent(
      mockGroupMany.memberSize + " members",
    );

    expect(singleQueryBtn).toHaveTextContent(
      mockGroupWithSingleQuery.querySize + " query",
    );

    expect(multiQueryBtn).toHaveTextContent(
      mockGroupMany.querySize + " queries",
    );

    expect(document.body).toMatchSnapshot();
    jest.restoreAllMocks();
  });
});
