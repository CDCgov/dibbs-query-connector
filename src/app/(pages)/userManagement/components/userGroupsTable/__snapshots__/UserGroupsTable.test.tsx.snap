// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`User Groups table renders the correct label syntax based on member/query size 1`] = `
<body>
  <div>
    <div
      class="userGroupsTable customizeTableWrapper"
    >
      <table
        class="usa-table usa-table--borderless customizeTable"
        data-testid="table"
      >
        <thead>
          <tr>
            <th
              class="grid-col-5"
            >
              Name
            </th>
            <th
              class="grid-col-3"
            >
              Members
            </th>
            <th
              class="grid-col-4"
            >
              Assigned queries
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="456"
            tabindex="0"
          >
            <td>
              <div
                class="buttonHoverGroup"
              >
                <span>
                  Order of the Phoenix
                </span>
                <button
                  class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-456"
                    >
                      Rename
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td>
              <button
                aria-description="Edit Order of the Phoenix members"
                class="usa-button usa-button--unstyled drawerButton margin-right-2 text-no-underline"
                data-testid="edit-member-list-0"
                type="button"
              >
                3 members
              </button>
            </td>
            <td
              class="grid-col-2"
            >
              <div
                class="buttonHoverGroup"
              >
                <button
                  aria-description="Edit Order of the Phoenix queries"
                  class="usa-button usa-button--unstyled drawerButton text-no-underline"
                  data-testid="edit-query-list-0"
                  type="button"
                >
                  3 queries
                </button>
                <button
                  class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline destructive-primary"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Trash icon indicating deletion of disease"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                    >
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
          <tr
            id="789"
            tabindex="0"
          >
            <td>
              <div
                class="buttonHoverGroup"
              >
                <span>
                  S.P.E.W.
                </span>
                <button
                  class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-789"
                    >
                      Rename
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td>
              <button
                aria-description="Edit S.P.E.W. members"
                class="usa-button usa-button--unstyled drawerButton margin-right-2 text-no-underline"
                data-testid="edit-member-list-1"
                type="button"
              >
                1 member
              </button>
            </td>
            <td
              class="grid-col-2"
            >
              <div
                class="buttonHoverGroup"
              >
                <button
                  aria-description="Edit S.P.E.W. queries"
                  class="usa-button usa-button--unstyled drawerButton text-no-underline"
                  data-testid="edit-query-list-1"
                  type="button"
                >
                  1 query
                </button>
                <button
                  class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline destructive-primary"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Trash icon indicating deletion of disease"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                    >
                      Delete
                    </span>
                  </span>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
